icefoganalytics/yhsi
# Issue 30: Site Policy Scoping Restrictions

Fixes https://github.com/icefoganalytics/yhsi/issues/30

# User Story
When a user goes to the /sites page, they can only see the sites that they are permitted to see.
When a user goes to an individual site (by direct URL), if they are not authorized to access the site, they are bounced back to the /sites page.

# Details
As see in https://github.com/icefoganalytics/yhsi/commit/17e03c861378e4996b2e4f3574632e3794468211 GET /place (sites list) is scoped by Security.SiteAccessType.
GET /place/:id is scoped by this as well.

# Implementation Details
- Moved get/search place router functions to a controller class for easier testing.
- Added a PlacePolicy for restricting access to a single place entity.
- Added a PlacePolicyScope for restricting access to a collection of places. 
- Move the User, UserSiteAccess, SiteAccessType models to appropriate model files and shim out attributes for standard JS casing.
- Add a bunch of development libraries for easier testing.

Core wiring looks like:
```js
// scoping for a single record
const policy = new PlacePolicy(currentUser, place);
if (policy.show()) {
  return res.json({
    data: place,
    relationships,
  });
}

return res.status(403).json({
  messages: [
    {
      variant: 'error',
      text: 'You are not authorized to access this content.',
    },
  ],
});

// scoping for collection of place objects
const allPlaces = db('Place');
const placePolicyScope = new PlacePolicyScope(allPlaces, currentUser);
const permittedPlaces = placePolicyScope.resolve();

return placeService
  .doSearch(permittedPlaces, query, sort, page, itemsPerPage, skip, take)
  .then((results) => {
    return res.json(results);
  })
```

# Screenshots
Viewing all records as an Adminstrator
![image](https://user-images.githubusercontent.com/23045206/164555712-2ccb25c9-e2cf-42b4-972f-174743abadaf.png)

Scoped by NTSMapSheet value
![image](https://user-images.githubusercontent.com/23045206/164557505-9809adb2-a376-436e-9372-338f76de84dd.png)


# Testing Instructions
1. Add the `Administrator` role to your user via the database console.
2. Go to the /sites route and you will see all 3842 site records.
3. Remove all roles from your user except the `Site Editor` role via the database console.
4. Add a new UserSiteAccess record that looks like `user_id=98, accest_type_id=1, access_text=115O/15`.
> Make sure you don't have any additional UserSiteAccess records.
5. Go to the /sites route again (after a hard refresh), you will now only see sites with an NTSMapSheet value of `115O/15`
